picoquic/picoquic.h

```diff
@@ -449,13 +450,47 @@ int picoquic_prepare_packet(picoquic_cnx_t* cnx,
     uint64_t current_time, uint8_t* send_buffer, size_t send_buffer_max, size_t* send_length,
     struct sockaddr_storage * p_addr_to, int * to_len, struct sockaddr_storage * p_addr_from, int * from_len);
 
-/* send and receive data on streams */
+/* Mark stream as active, or not.
+ * If a stream is active, it will be polled for data when the transport
+ * is ready to send.
+ * Returns an error if data was previously queued on the stream using
+ * "picoquic_add_to_stream" and the data has not been sent yet.
```

Comment:
I do not love "Returns an error if data was previously queued on the stream using 'picoquic_add_to_stream' and the data has not been sent yet".

The interface seems to allow the application to sometimes use picoquic_add_to_stream and sometimes use the callback functionality.  However, the application cannot reliably use callbacks after using picoquic_add_to_stream, because it does not known when all data queued via picoquic_add_to_stream has been sent.

The cleaner semantics would allow picoquic_mark_active_stream any time.  When the transport sends all previously-buffered data, if the steam is marked active, the application callback will be called to get more data.  It is important that picoquic_add_to_stream cancels the "active" status (as it does now).

I'd propose this:

picoquic_mark_active_stream IS allowed while there is buffered data.  The library implementation is something like:

```
if stream is active:
    callback_max_bytes = max_bytes_for_stream - num_buffered_bytes
    if callback_max_bytes > 0:
          call_callback(..., callback_max_bytes, .... )
     else
           write_buffered_bytes_to_stream_frame(min(max_bytes_for_stream, num_buffered_bytes))
```
picoquic_provide_stream_data_buffer implementation:
```
    data_buf_ptr = write_stream_frame_header(num_buffered_bytes + nb_bytes)
    if buffered_bytes > 0
         copy_buffered_bytes(data_buf_ptr, num_buffered_bytes)
    return data_buf_ptr + num_buffered_bytes
```

A solid implementation will also make sure to disallow picoquic_add_to_stream while in the callback (via a mutex, for example -- deadlock if called from the callback thread itself; wait if multiple threads).

Reply:
You are right, it is not hard to allow active marking at any time. And in fact it makes sense. For example, a web server might queue a header frame on the stream, and then rely on the callback to feed the frame's data.

The relation between application and transport is more "intimate" with an in-process transport like QUIC than with a kernel transport like TCP. If the application process crashes, the transport also crashes. The priority cannot be to try for full isolation -- we cannot. Instead, I would rather try for easy debugging: it should be easy for the app developer to understand why things are breaking, and to fix the application. The priority should thus be, return good error messages, plus have a reasonable documentation. But I guess that creating a deadlock would be a good way to provide feedback to the developer...